tests/test-cmd/test.build.js

```diff
@@ -37,6 +37,40 @@ describe('build', () => {
     );
   });
 
+  it('zips a package with custom file extension', () => {
+    return withTempDir(
+      (tmpDir) =>
+        build({
+          sourceDir: fixturePath('minimal-web-ext'),
+          artifactsDir: tmpDir.path(),
+          extension: 'blah',
+        })
+        .then((buildResult) => {
+          assert.match(buildResult.extensionPath,
+                       /minimal_extension-1\.0\.blah$/);
+          return buildResult.extensionPath;
+        })
+    );
+  });
+
+  it('throws an error when invalid extension is provided', () => {
+    let failed = false;
+    try {
```

Comment:
all callers will expect to treat `build()`'s return value as a promise so I think it's time to wrap the first part of that function in a promise:
````javascript
return new Promise(
  (resolve) => {
    if (!allowedExtension.test(fileExtension)) {
      throw new Error('...');
    }
    resolve();
  })
  .then(() => prepareArtifactsDir(artifactsDir));
````

@rpl is embarking on an epic patch to switch us to async/await! https://github.com/mozilla/web-ext/pull/411